### ISSUE TITLE
Question: why character grid?

### ISSUE THREAD

**Issue by maximbaz** [kovidgoyal/kitty#704]
This is not a bug or a feature request, but a question that I'm genuinely interested to ask you.

This project's goal is to make a terminal that is modern and sexy, not be restricted by limitations of the past age. GPU rendering, graphics, unicode, true color, ligatures, window layouts and much more, everything is very pleasant. 

So why are we still bound by the character grid? Isn't this just another artifact of the age of hardware terminals? 

I managed to find [this comment](https://github.com/kovidgoyal/kitty/issues/523#issuecomment-386850678) saying that performance would have been terrible otherwise, but is that not an exaggerated statement? I'm sure it is possible to write GUI apps that are maybe slower, but still not terrible at performance.

But I wonder if there is more to it. Some feature requests were closed as "this cannot be done in grid-based terminals", I want to understand if having character grid in the first place was a conscious decision, and if so what were the reasons behind it.

**Comment by kovidgoyal** [comment]
Because this is a terminal emulator. And terminals are character based devices. Limitations are often a spur for creativity and the limitation of needing to interact via character based devices has resulted in an entire ecosystem of terminal programs whose interface and design characteristics are both limited to and also enhanced by the fundamental nature of the TTY device. This is not a set of constraints I want to lift. So my mission is to make the best possible *terminal emulator*.

And just incidentally, when I say performance, I dont mean raw rendering performance. Even 3D games have more than adequate raw rendering performance on modern hardware. Most modern termial emulators have more than adequate performance. 

By performance, I mean CPU usage to perform a task.See https://sw.kovidgoyal.net/kitty/performance.html

**Comment by maximbaz** [comment]
I see, thanks for the answer. So do I understand correctly that if there was an app that looked like kitty, but wasn't based on a grid and allowed variable-width fonts, you wouldn't call that app a "terminal emulator"? Also, is it just a game of definitions, or there would be some conceptual challenges, e.g. vim wouldn't be able to run if there was no fixed number of columns in a row? Sorry if that's a dumb question, but I don't know if vim actually depends on a fact that there is the same number of columns in all rows.

Thanks for explaining the meaning of performance in this context, quite interesting.

**Comment by Althorion** [comment]
AFAIK (I don’t have the actual font) Pragmata Pro _is_ monospaced, just it implements ligatures differently than Fira Code—every simple glyph is the same size and ligatures are a multiple of that.

The same, I believe, goes for [Iosevka](https://github.com/be5invis/Iosevka), which doesn’t work in Kitty either, but never caused me any problems in all other fix-width only applications.

**Comment by kovidgoyal** [comment]
According to #557 pragmata pro's ligatures work fine in kitty. As for Iosevka it insanely supports infinite length ligatures, which I absolutely refuse to do for performance reasons, see #297

**Comment by ahmedelgabri** [comment]
> According to https://github.com/kovidgoyal/kitty/issues/557 pragmata pro's ligatures work fine in kitty. As for Iosevka it insanely supports infinite length ligatures, which I absolutely refuse to do for performance reasons, see https://github.com/kovidgoyal/kitty/issues/297

Font choices are a personal preference & it depends on many things, what the person likes, their screen, eye sight, OS rendering, etc... & also not everyone can afford PragmataPro. Plus people expect fonts to _just work_, especially if they work in all other apps on their system just fine. 

Asking people to change what they like shouldn't be part of installing an app, it's like releasing an app with only one theme & asking everyone else to just get used to this theme, won't work. Because these things are very subjective & depends on taste & other factors. 

Kitty is by far the best terminal emulator right now & I really appreciate the work done in this project. This is the only thing that it's missing to be complete IMO. I really hope that this can be changed at some point.

This is just my 2 cents.

**Comment by andreypopp** [comment]
@ahmedelgabri maybe Iosevka should be fixed instead?

**Comment by Althorion** [comment]
I’ve [asked](https://github.com/be5invis/Iosevka/issues/285) for a compatible version to be pre-build and you can build one yourself from the source, I think—my nodejs installation hates me, so I can’t just go and check by myself, will try to find some time to fix it and rebuild.

**Comment by ahmedelgabri** [comment]
@andreypopp Both sides have valid reasons in the argument, I just hope we can find a middle ground. See Iosevka's author comments https://github.com/kovidgoyal/kitty/issues/297#issuecomment-360972776 & https://github.com/kovidgoyal/kitty/issues/297#issuecomment-360980828

**Comment by kovidgoyal** [comment]
I'm afraid there is no middle ground, at least that I can see. As long as Iosevka insists on having infinite length ligatures, there is simply no way to support it in kitty. It's a pity, but that's what it is.

And I never suggested using Pragmata Pro, I simply mentioned that unlike Iosevka, Pragmata Pro's ligatures do work in kitty. As do the ligatures from many other fonts. Indeed, I think Iosevka is the only one that does not. And even with Iosevka if you really, really want to use it, use the variant without ligatures.

**Comment by fuloating** [comment]
I am not sure if this is a related question, or whether I should have created a new issue or contacted through other means, but I want to know if there is a solution in kitty for this scenario:

I created a test font in fontforge from Monoid font, renamed its family to Monoid2. What I did to it is I created a ligature substitution in the lookup to substitute "ab" to "¡". Now when I use this font in the qutebrowser, the string "abc" looks like "¡c" and same is with another terminal emulator qterminal, but with kitty it looks like "¡ c". It reserves an extra block, and what I understood from the technical text in some of the discussions here it is a feature.

My question is if there is a way to make it appear like "¡c" in anyway in kitty?
